cdc/sink/common/flow_control.go

```diff
@@ -0,0 +1,209 @@
+// Copyright 2021 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package common
+
+import (
+	"log"
+	"sync"
+	"sync/atomic"
+
+	"github.com/edwingeng/deque"
+	"github.com/pingcap/errors"
+	cerrors "github.com/pingcap/ticdc/pkg/errors"
+	"go.uber.org/zap"
+)
+
+// TableMemorySizeController is designed to curb the total memory consumption of processing
+// the event streams in a table.
+// A higher-level controller more suitable for direct use by the processor is TableFlowController.
+type TableMemorySizeController struct {
+	Quota uint64 // should not be changed once intialized
+
+	IsAborted uint32
+
+	mu       sync.Mutex
+	Consumed uint64
+
+	cond *sync.Cond
+}
+
+// NewTableMemorySizeController creates a new TableMemorySizeController
+// quota: max advised memory consumption in bytes.
+func NewTableMemorySizeController(quota uint64) *TableMemorySizeController {
+	ret := &TableMemorySizeController{
+		Quota:    quota,
+		mu:       sync.Mutex{},
+		Consumed: 0,
+	}
+
+	ret.cond = sync.NewCond(&ret.mu)
+	return ret
+}
+
+// ConsumeWithBlocking is called when a hard-limit is needed. The method will
+// block until enough memory has been freed up by Release.
+// blockCallBack will be called if the function will block.
+// Should be used with care to prevent deadlock.
+func (c *TableMemorySizeController) ConsumeWithBlocking(nBytes uint64, blockCallBack func() error) error {
+	if nBytes >= c.Quota {
+		return cerrors.ErrFlowControllerEventLargerThanQuota.GenWithStackByArgs(nBytes, c.Quota)
+	}
+
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	calledBack := false
+	for {
+		if atomic.LoadUint32(&c.IsAborted) == 1 {
+			return cerrors.ErrFlowControllerAborted.GenWithStackByArgs()
+		}
+		if c.Consumed+nBytes < c.Quota {
+			break
+		}
+
+		if !calledBack {
+			calledBack = true
+			err := blockCallBack()
+			if err != nil {
+				return errors.Trace(err)
+			}
+		}
+		c.cond.Wait()
+	}
+
+	c.Consumed += nBytes
+	return nil
+}
+
+// ForceConsume is called when blocking is not acceptable and the limit can be violated
+// for the sake of avoid deadlock. It merely records the increased memory consumption.
+func (c *TableMemorySizeController) ForceConsume(nBytes uint64) error {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	if atomic.LoadUint32(&c.IsAborted) == 1 {
+		return cerrors.ErrFlowControllerAborted.GenWithStackByArgs()
+	}
+
+	c.Consumed += nBytes
+	return nil
+}
+
+// Release is called when a chuck of memory is done being used.
+func (c *TableMemorySizeController) Release(nBytes uint64) {
+	c.mu.Lock()
+
```

Comment:
how about `defer c.mu.Unlock()` at this place ?

Reply:
It would be correct implementation, but it will likely cause a block consumer to be waken up only to be blocked again immediately by `mu` that is still locked.